standards/core/rln-contract.md

```diff
@@ -0,0 +1,204 @@
+
+- the overall smart contract functionality;
+- contract parameters;
+- the values of those parameters for the initial deployment;
+- the way in which the parameters can be modified.
+
+## Background
+
+Rate-Limiting Nullifiers (RLN) is a ZK-based rate-limiting technique used in Waku.
+Before sending messages, users must register in a smart contract.
+They then attach a ZK proof of valid membership to every message.
+Relaying nodes check message validity and drop invalid messages.
+
+A message is considered valid in terms of RLN if:
+- its sender has an active RLN membership; and
+- the sender hasn't exceeded their rate limit within the current epoch.
+
+RLN is managed by a smart contract that keeps track of active memberships.
+Users interact with the contract to register a new membership and
+to obtain Merkle proofs and the tree root,
+which is necessary for proof generation and verification.
+
+In testnet deployment, the RLN contract does not require any payment apart from gas costs.
+This document aims to make the RLN contract suitable for mainnet deployment.
+In particular, the contract would require uses to put down a deposit.
+The aim of the deposit initially is purely to protect the system from abuse.
+It may also be explored as a revenue stream for Waku in later versions.
+
+## Semantics
+
+The RLN contract provides the following functionality:
+- register a membership in a given tier;
+- check if a membership is valid;
+- withdraw the deposit for an expired membership;
+- extend an expired membership.
+
+For the `Owner`, the contract provides the following additional functionality:
+- change the modifiable parameters (see parameter table below);
+- (TBD) freeze certain functionality (e.g., in case of an attack, stop new registrations, deposit withdrawals, etc).
+
+In the initial deployment:
+- there is no slashing;
+- membership expiration includes a grace period;
+- a user can extend their membership without putting up any additional deposit.
+
+### Membership life-cycle
+
+To register a membership, a user puts locks up a deposit.
+Each membership has an expiration date.
+After a membership expires, it enters a grace period, during which the user can extend it.
+Extending a membership requires no additional deposit.
+After the grace period, an expired membership can be taken over by another user.
+At any point, a user can withdraw their deposit and terminate the membership.
+
+In more detail, the membership life-cycle is as follows:
+1. A user registers a membership:
+    1. If there are expired unclaimed slots in the tree, use the slot with the earliest expiration date past its grace period, otherwise create a new slot;
+    2. if the membership re-uses a previously used slot, refund the previous user’s deposit.
+2. The user sends messages while the membership is active:
+    1. If the user exceeds the rate limit, their over-limit messages are not propagated in that epoch; the deposit is not slashed.
```

Comment:
> are not propagated in that epoch

This made me think about whether they would be propagated in the next epoch somehow... So maybe change the wording to mention that the messages get dropped/ignored by the nodes?

Reply:
> whether they would be propagated in the next epoch somehow.

Isn't every message committed to one epoch? My understanding was that a message sent in the wrong epoch would be dropped because of that reason, regardless of proof validity - is this true?

Reply:
> Isn't every message committed to one epoch?

AFAIU now, this is the case. Removed "within this epoch": the message won't be propagated within next epochs either simply because it's commited to a wrong epoch by then.